core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -25,34 +37,275 @@ type GCSFileTransfer struct {
+		}
+		if objAttrs.Etag != task.Digest {
+			err := fmt.Errorf(
+				"digest/etag mismatch: etag %s does not match expected digest %s",
+				objAttrs.Etag,
+				task.Digest,
+			)
+			return formatDownloadError("", err)
+		}
+		return ft.downloadFile(object, task.Path)
+	}
+	return nil
+}
+
+// DownloadFile downloads the contents of object into a file at path localPath.
+func (ft *GCSFileTransfer) downloadFile(
+	object *storage.ObjectHandle,
+	localPath string,
+) (err error) {
+	r, err := object.NewReader(ft.ctx)
+	if err != nil {
+		return formatDownloadError("error creating reader", err)
+	}
+	defer r.Close()
+
+	// Check if the directory exists, and create it if it doesn't
+	dir := path.Dir(localPath)
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error creating download directory %s", dir),
+			err,
+		)
+	}
+
+	// open the file for writing and defer closing it
+	file, err := os.Create(localPath)
+	if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error creating download file %s", localPath),
+			err,
+		)
+	}
+
+	defer func(file *os.File) {
+		fileError := file.Close()
+		if err == nil {
+			err = formatDownloadError(
+				fmt.Sprintf("error closing file %s", localPath),
+				fileError,
+			)
+		}
+	}(file)
+
+	_, err = io.Copy(file, r)
+	if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error copying file %s", localPath),
+			err,
+		)
+	}
+
+	return nil
+}
+
+// getObjectAndAttrs returns the object handle and attributes if the object
+// exists and corresponds with a file.
+//
+// If the object corresponds to a directory, we return the ErrObjectIsDirectory
+// error to skip downloading without failing.
+func (ft *GCSFileTransfer) getObjectAndAttrs(
+	bucket *storage.BucketHandle,
+	task *ReferenceArtifactDownloadTask,
+	key string,
+) (*storage.ObjectHandle, *storage.ObjectAttrs, error) {
+	object, err := ft.getObject(bucket, task, key)
+	if err != nil {
+		return nil, nil, err
+	}
+	if strings.HasSuffix(object.ObjectName(), "/") {
+		return nil, nil, ErrObjectIsDirectory
+	}
+
+	attrs, err := object.Attrs(ft.ctx)
+	// if object doesn't have an extension and has size 0, it might be a
+	// directory as we cut off the ending "/" in the manifest entry.
+	if err != nil && (path.Ext(object.ObjectName()) == "" && task.Size == 0) {
+		key := key + "/"
+		object, err := ft.getObject(bucket, task, key)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// if the new key corresponds with an existing directory, it should
+		// return the attrs without an error
+		_, err = object.Attrs(ft.ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		return nil, nil, ErrObjectIsDirectory
+	} else if err != nil {
+		return nil, nil, err
+	}
+	return object, attrs, nil
+}
+
+// getObject returns the object handle given a bucket, key, and task object
+// with a possible versionId.
+func (ft *GCSFileTransfer) getObject(
```

Comment:
Question: I'm assuming if there is no version given, with a path without a version it returns the `latest` object?

Reply:
objects in GCS should actually always have a version, but if for some reason _we_ didn't store that version id, it would just find the latest object yes.

side note: our GCS download implementation in Python doesn't search for the version that has a matching digest (if the latest one doesn't match the manifest), so I also left that out and just errored out later- but if we wanted to add that logic in it would be in this section